package com.joaodev.taskmanager.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

public abstract class TaskRepositoryTest {

    TaskRepository repository;

    protected abstract TaskRepository createRepository();

    @BeforeEach
    void setUp() {
        repository = createRepository();
    }

    @Test
    void should_save_and_retrieve_task_by_id() {
        // given
        var task = new Task("Passar na padaria", Optional.empty());

        // when
        var saved = repository.save(task);
        Optional<Task> result = repository.findById(saved.getId());

        // then
        assertThat(result).isPresent();
        assertThat(result.get().getId()).isEqualTo(saved.getId());
        assertThat(result.get().getTitle()).isEqualTo(saved.getTitle());
        assertThat(result.get().getDescription()).isEqualTo(saved.getDescription());
    }

    @Test
    void should_find_all_persisted_tasks() {
        // given
        var task1 = new Task("Arrumar chuveiro", Optional.of("Comprar chuveiro novo"));
        var task2 = new Task("Trocar interruptor", Optional.of("Encontrar chave de fenda"));

        repository.save(task1);
        repository.save(task2);

        // when
        List<Task> tasks = repository.findAll();

        // then
        assertThat(tasks).hasSize(2);
        assertThat(tasks).extracting(Task::getId).containsExactlyInAnyOrder(task1.getId(), task2.getId());
        assertThat(tasks).extracting(Task::getTitle).containsExactlyInAnyOrder(task1.getTitle(), task2.getTitle());
        assertThat(tasks).extracting(Task::getDescription).containsExactlyInAnyOrder(task1.getDescription(), task2.getDescription());
    }

    @Test
    void should_delete_tasks_by_id() {
        // given
        var task = repository.save(new Task("Treinar na academia", Optional.of("Levar água")));
        var taskId = task.getId();

        // when
        repository.delete(taskId);
        Optional<Task> result = repository.findById(taskId);

        // then
        assertThat(result).isEmpty();
    }

    @Test
    void should_return_empty_when_searching_non_existent_task() {
        // given
        var nonExistingId = new TaskId();

        // when
        Optional<Task> result = repository.findById(nonExistingId);

        // then
        assertThat(result).isEmpty();
    }

    @Test
    void should_update_task_status_successfully() {
        // given
        var task = repository.save(new Task("Atualizar Carteira de Habilitação", Optional.empty()));
        task.setDescription(Optional.of("Não expirou ainda"));
        task.setStatus(TaskStatus.IN_PROGRESS);

        // when
        repository.save(task);
        Optional<Task> result = repository.findById(task.getId());

        // then
        assertThat(result).isPresent();
        assertThat(result.get().getDescription()).isEqualTo(Optional.of("Não expirou ainda"));
        assertThat(result.get().getStatus()).isEqualTo(TaskStatus.IN_PROGRESS);
    }
}